File: Game/src/main/java/net/winternetwork/bedwars/game/module/stage/model/ToStartStage.java
package net.winternetwork.bedwars.game.module.stage.model;

import lombok.Getter;
import net.winternetwork.bedwars.api.game.flag.Flag;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.api.score.ReplaceableList;
import net.winternetwork.bedwars.game.module.Modules;
import net.winternetwork.bedwars.game.module.map.MapModule;
import net.winternetwork.bedwars.game.module.map.object.GameMap;
import net.winternetwork.bedwars.game.module.stage.StageManager;
import net.winternetwork.bedwars.game.module.stage.StageModule;
import net.winternetwork.bedwars.game.module.stage.flag.Flags;
import net.winternetwork.bedwars.game.settings.GameSettings;
import org.bukkit.entity.Player;

import java.util.Arrays;
import java.util.List;

public class ToStartStage extends Stage {

    @Getter(lazy = true)
    private final StageManager stageManager = Modules.getModule(StageModule.class).getStageManager();
    private boolean started = false;

    public ToStartStage() {
        super("Começando.", 20);
    }

    @Override
    public List<Flag> getFlags() {
        return Arrays.asList(
                Flags.NO_PVP,
                Flags.NO_BREAK,
                Flags.NO_BUILD,
                Flags.NO_GENERATOR
        );
    }

    @Override
    public ReplaceableList getScoreboard() {
        return new ReplaceableList(
                Arrays.asList(
                        "&1",
                        "&fMapa: &a<mapa>",
                        "&fComeçando em: &a<next>",
                        "&2",
                        "&fJogadores: &a<online>/<max>",
                        "&3",
                        "&ewww.baianowars.kt"
                ),
                (s, player) -> s
                        .replaceAll("<mapa>", "Gadolandia")
                        .replaceAll("<online>", getOnlinePlayers() + "")
                        .replaceAll("<max>", "15")
                        .replaceAll("<next>", String.format("%ss", getTimeLeft()))
        );
    }

    @Override
    public void onStageJoin() {
    }

    @Override
    public void onSecondPassed() {

        if (getTimeLeft() <= 0) {
            getStageManager().callNextStage();
        }

        if (getTimeLeft() <= 5) {
            broadcast(
                    String.format("Começando o jogo em %ss.", getTimeLeft())
            );
        }
    }

    @Override
    public void onTimeLeft() {
        if (getOnlinePlayers() < GameSettings.PLAYERS_TO_START) {
            broadcast("§eAinda não tem jogadores suficientes para começar.");

            getStageManager().callPreviousStage();
            return;
        }

        started = true;
        getStageManager().callNextStage();
    }

    @Override
    public void onStageExit() {
        if (!started) return;
        broadcast("Começou!");

        GameMap map = Modules.getModule(MapModule.class).getMapManager().getAll().get(0);

        System.out.println(map.toString());

        for (int i = 0; i < getAllPlayers().size(); i++) {
            Player player = getAllPlayers().get(i);

            player.teleport(map.getSpawnLocations()[i]);
        }
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/game/modules/stage/object/ToStartStage.kt
package net.winternetwork.bedwars.game.modules.stage.`object`

import net.winternetwork.bedwars.api.game.stage.Stage
import net.winternetwork.bedwars.api.score.ReplaceableList
import net.winternetwork.bedwars.game.modules.maps.MapModule
import net.winternetwork.bedwars.game.modules.stage.StageModule
import net.winternetwork.bedwars.game.modules.stage.flag.Flags
import net.winternetwork.bedwars.game.settings.GameSettings

class ToStartStage(
        stageModule: StageModule,
        mapModule: MapModule
) : Stage("Starting", 20) {

    private val stageManager = stageModule.stageManager

    private val mapManager = mapModule.mapManager

    private var started = false

    override val flags = listOf(
            Flags.NO_PVP_FLAG,
            Flags.NO_BREAK_FLAG,
            Flags.NO_BUILD_FLAG,
            Flags.NO_GENERATOR_FLAG
    )

    override val scoreboard = ReplaceableList(
            listOf(
                    "&1",
                    "&fMapa: &a<mapa>",
                    "&fComeçando em: &a<next>",
                    "&2",
                    "&fJogadores: &a<online>/<max>",
                    "&3",
                    "&ewww.baianowars.kt"
            )
    ) { s, player ->
        s
                .replace("<mapa>", "Gadolandia")
                .replace("<online>", "$onlinePlayers")
                .replace("<max>", "${GameSettings.MAX_PLAYERS}")
                .replace("<next>", "${timeLeft}s")
    }

    override fun onSecondPassed() {
        when {
            timeLeft <= 0 -> {
                stageManager!!.next()
                return
            }
            timeLeft <= 5 -> broadcast("§eComeçando o jogo em ${timeLeft}s")
        }
    }

    override fun onTimeLeft() {
        if (onlinePlayers < GameSettings.PLAYERS_TO_START) {
            broadcast(
                    "§eAinda não tem jogadores suficientes para começar."
            )
            stageManager!!.previous()
            return
        }

        started = true
        stageManager!!.next()
    }

    override fun onStageExit() {
        if (!started) return
        broadcast("§2Começou!")

        val map = mapManager.all()[0]

        allPlayers.forEachIndexed { i, p ->
            p.teleport(map.locations[i])
        }
    }

}